core/src/avm2/globals/flash/xml/XMLDocument.as

```diff
@@ -1,21 +1,84 @@
+
+import flash.xml.XMLNode;
+import flash.xml.XMLNodeType;
+
    public class XMLDocument extends XMLNode {
 
       public var ignoreWhite: Boolean = false;
 
       public function XMLDocument(input: String = null) {
-         super(0, null);
-         _isDocument = true;
-         parseXML(input);
+         super(XMLNodeType.ELEMENT_NODE, null);
+         if (input != null) {
+            parseXML(input);
+         }
       }
 
       public function parseXML(input: String): void {
-         // Unlike the new AS3 XML class, XMLDocument allows parsing
-         // multiple elements like <a>1</a><b>2</b>. So introduce a wrapper.
-         _setXML("<xml>" + input + "</xml>", this.ignoreWhite);
-         _children = null;
+         // This is something of a hack, but that's somewhat the nature of XMLDocument
+         // It accepts things like `<node>...</node> <!-- comment --> foo` which is FOUR children:
+         // `<node>...</node>` gets parsed as an element
+         // ` ` gets parsed as a text node
+         // `<!-- comment -->` gets parsed as a comment
+         // ` foo` is another text node
+         // To achieve this, just wrap it all together in a parent.
+
+         var oldSettings = XML.AS3::settings();
+         var newSettings = XML.AS3::defaultSettings();
+         newSettings.ignoreWhitespace = this.ignoreWhite;
+         XML.AS3::setSettings(newSettings);
+
+         try {
+            clear();
+            var root = new XML("<xml>" + input + "</xml>");
+            for each (var child in root.children()) {
+               var node = _convertXmlNode(child);
+               if (node != null) {
+                  appendChild(node);
+               }
+            }
+         } finally {
+            XML.AS3::setSettings(oldSettings);
+         }
+      }
+
+      private function _convertXmlNode(original: XML): XMLNode {
+         var nodeType = _convertXmlNodeType(original.nodeKind());
+         if (nodeType == 0) {
+            return null;
+         }
+         var nodeValue = nodeType == XMLNodeType.ELEMENT_NODE ? original.name() : original.toString();
+         var result = new XMLNode(nodeType, nodeValue);
+         for each (var originalChild in original.children()) {
+            var child = _convertXmlNode(originalChild);
+            if (child != null) {
+               result.appendChild(child);
+            }
+         }
+         var attributeList = original.attributes();
+         var attributes = {};
+         for each (var attribute in attributeList) {
+            attributes[attribute.name()] = attribute.toString();
+         }
+         result.attributes = attributes;
+         return result;
+      }
+
+      private function _convertXmlNodeType(kind: String): uint {
+         if (kind == "text") {
+            return XMLNodeType.TEXT_NODE;
+         }
+         if (kind == "comment") {
+            return XMLNodeType.COMMENT_NODE;
+         }
+         if (kind == "processing-instruction") {
```

Comment:
I don't think this is possible. We don't allow processing-instructions during parsing. 

In general: I would  suggest we just throw for what I think is the impossible situation of an unexpected kind here. (Would also eliminate some null checks above)

Reply:
We don't exactly have a way to panic or throw RustErrors during AS, and throwing a new type of AS error is weird and unprecedented.

Reply:
I don't think it's unreasonable to do `throw new Error("unexpected state")` for something that we believe is impossible. This seems a lot easier to debug later than the number of children being different or something like that.

Reply:
Could you add in a stub, so that there's some indication if this ever gets hit?

Reply:
Stub will incorrectly mark this as a stub, which it isn't. I'll add an error.